flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -81,16 +92,12 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
 				Object old = clients.putIfAbsent(connectionId, connectingChannel);
 
 				if (old == null) {
```

Comment:
TBH I do not like the previous way of implementing this logic via `ConcurrentHashMap`, because it results in five different `if..else` code paths together and some paths are almost very similar. 

Instead if we realize it via an explicit `synchronized` lock outside, then the internal logic seems very simple, because we do not need to consider the race conditions of connect/wait among multiple threads.  Only one thread would actually establish the connection and wait, all the other threads would be blocked outside of `synchronized` until the connection established or exception after retries.

It seems not get any benefits to have the small scope lock via `ConcurrentHashMap`, because this operation will be done only once, not frequent operations. But I am not forcing to refactor it in this PR since it can be considered as another issue.

Reply:
As the race condition only happens among the subpartitions for the same channel, the syncronized lock is probably too heavy. If considering a job with 256 tms, 255 connection needs to build. The lock will introduce a longer job initialization time.